Sources/Core/GTMSessionFetcher.m

```diff
@@ -1972,7 +1972,19 @@ - (void)stopFetching {
     // Prevent enqueued callbacks from executing.
     _userStoppedFetching = YES;
   }  // @synchronized(self)
-  [self stopFetchReleasingCallbacks:YES];
+
+  if (self.stopFetchingTriggersCallbacks) {
+    NSError *error = [NSError errorWithDomain:kGTMSessionFetcherErrorDomain
+                                         code:GTMSessionFetcherErrorCancelled
+                                     userInfo:@{@"description" : @"Operation cancelled"}];
+    [self invokeFetchCallbacksOnCallbackQueueWithData:nil
+                                                error:error
+                                          mayDecorate:YES
+                               shouldReleaseCallbacks:YES];
+    [self sendStopNotificationIfNeeded];
```

Comment:
This doesn't trigger stopping the fetch, just invokes the callback and sends the notification.

I think the changes needs to be spread across `stopFetchingReleasingCallbacks:` and `URLSession:task:didCompleteWithError:`.

Here, instead of this code we pass `NO` if `self.stopFetchingTriggersCallbacks` to this initial `stopFetchingReleasingCallbacks:` and in `URLSession:task:didCompleteWithError:` checks for an error with `NSURLErrorDomain` and code `NSURLErrorCancelled` and perhaps wraps that error before calling the callback.

WDYT?

Reply:
I couldn't get the changes to work in `stopFetchingReleasingCallbacks` because of other call paths into it.   Instead, I reverted to always calling  `stopFetchingReleasingCallbacks`  from `stopFetching`. 

I'm also not sure what to do in `URLSession:task:didCompleteWithError:`. Based on https://github.com/google/gtm-session-fetcher/blob/main/UnitTests/GTMSessionFetcherFetchingTest.m#L886, it looks like callbacks occur if the cancellation happens on the server.

Reply:
My worry is that this is a new codepath to firing the completion handler, and I'm not sure the implications of firing the handler here before the rest of the flow has completed.

My suggestion is that we allow the task cancellation to take place, but in `finishWithError:shouldRetry:` we would want to check for a cancellation error from NSURLSession, and if so check whether `_userStoppedFetching` is set, and let the callback flow happen that way. Would also need to ensure the completion handler block will fire after user-stopped for that case.

When `cancel` is called on a task, `NSURLErrorCancelled` is returned: https://developer.apple.com/documentation/foundation/nsurlsessiontask/1411591-cancel